cpu/stm32_common/periph/flashpage.c

```diff
@@ -15,76 +16,176 @@
+            FLASH->PRGKEYR = FLASH_PRGKEY2;
+        }
+    }
+#endif
+}
 
-    /* the internal RC oscillator (HSI) must be enabled */
-    RCC->CR |= (RCC_CR_HSION);
-    while (!(RCC->CR & RCC_CR_HSIRDY)) {}
+static void _lock(void)
+{
+    DEBUG("[flashpage] locking the flash module\n");
+    CNTRL_REG |= CNTRL_REG_LOCK;
+}
 
-    /* unlock the flash module */
-    DEBUG("[flashpage] unlocking the flash module\n");
-    if (FLASH->CR & FLASH_CR_LOCK) {
-        FLASH->KEYR = FLASH_KEY1;
-        FLASH->KEYR = FLASH_KEY2;
-    }
+static void _erase_page(void *page_addr)
+{
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1)
+    uint32_t *dst = page_addr;
+#else
+    uint16_t *dst = page_addr;
+#endif
+
+   /* unlock the flash module */
+    _unlock();
 
-    /* ERASE sequence */
     /* make sure no flash operation is ongoing */
     DEBUG("[flashpage] erase: waiting for any operation to finish\n");
     while (FLASH->SR & FLASH_SR_BSY) {}
     /* set page erase bit and program page address */
-    DEBUG("[flashpage] erase: setting the erase bit and page address\n");
-    FLASH->CR |= FLASH_CR_PER;
-    FLASH->AR = (uint32_t)flashpage_addr(page);
-    DEBUG("address to erase: %p\n", flashpage_addr(page));
+    DEBUG("[flashpage] erase: setting the erase bit\n");
+    CNTRL_REG |= FLASH_CR_PER;
+    DEBUG("address to erase: %p\n", page_addr);
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1)
+    DEBUG("[flashpage] erase: trigger the page erase\n");
+    *dst = (uint32_t)0;
+#else
+    DEBUG("[flashpage] erase: setting the page address\n");
+    FLASH->AR = (uint32_t)dst;
     /* trigger the page erase and wait for it to be finished */
     DEBUG("[flashpage] erase: trigger the page erase\n");
-    FLASH->CR |= FLASH_CR_STRT;
+    CNTRL_REG |= FLASH_CR_STRT;
+#endif
     DEBUG("[flashpage] erase: wait as long as device is busy\n");
     while (FLASH->SR & FLASH_SR_BSY) {}
     /* reset PER bit */
     DEBUG("[flashpage] erase: resetting the page erase bit\n");
-    FLASH->CR &= ~(FLASH_CR_PER);
+    CNTRL_REG &= ~(FLASH_CR_PER);
+
+    /* lock the flash module again */
+    _lock();
+}
+
+void flashpage_write_raw(void *target_addr, void *data, size_t len)
+{
+    /* assert multiples of FLASHPAGE_RAW_BLOCKSIZE are written and no less of
+       that length. */
+    assert(!(len % FLASHPAGE_RAW_BLOCKSIZE));
+
+    /* ensure writes are aligned */
+    assert(!(((unsigned)target_addr % FLASHPAGE_RAW_ALIGNMENT) ||
+            ((unsigned)data % FLASHPAGE_RAW_ALIGNMENT)));
+
+    /* ensure the length doesn't exceed the actual flash size */
+    assert(((unsigned)target_addr + len) <
+           (CPU_FLASH_BASE + (FLASHPAGE_SIZE * FLASHPAGE_NUMOF)) + 1);
+
+#if defined(CPU_FAM_STM32L0) || defined(CPU_FAM_STM32L1)
+    uint32_t *dst = target_addr;
+    uint32_t *data_addr = (uint32_t *)data;
+#else
+    uint16_t *dst = (uint16_t *)target_addr;
+    uint16_t *data_addr = (uint16_t *)data;
+#endif
+
```

Comment:
HSI needs to be enabled here and disabled after locking, otherwise it crashes on stm32f091

Reply:
> HSI needs to be enabled here and disabled after locking

Thanks, I'll try that

Reply:
Thanks a lot @vincent-d, that was it! I refactored a bit the code again and tested on nucleo-l073 and nucleo-f070, it works on both.